src/main/java/org/skriptlang/skript/lang/properties/Property.java

```diff
@@ -0,0 +1,255 @@
+ * 1) Using {@link PropertyBaseExpression}. This expression handles all the complexities of properties for any property
+ * handler that implements {@link ExpressionPropertyHandler}. See {@link PropExprName} for an example.<br>
+ * 2) By implementing the property directly in a custom syntax. This is more complex, but allows for more flexibility.
+ * See {@link PropCondContains} for an example. The implementer is responsible for using {@link PropertyBaseSyntax#asProperty(Property, Expression)}
+ * and {@link PropertyBaseSyntax#getPossiblePropertyInfos(Property, Expression)} to ensure the given expression can return
+ * valid types that have the given property, and then use {@link PropertyMap#get(Class)} during runtime
+ * to acquire the right handler for the given type and then apply it.
+ * <br>
+ * <br>
+ * All properties should be registered with the {@link PropertyRegistry} before use, to ensure that no conflicts between
+ * registered properties occur.
+ *
+ * @param <Handler> the type of the handler for this property
+ */
+public record Property<Handler extends PropertyHandler<?>>(
+		String name,
+		String description,
+		String[] since,
+		SkriptAddon provider,
+		@NotNull Class<? extends Handler> handler
+) {
+
+	private static final PropertyRegistry PROPERTY_REGISTRY = Skript.getAddonInstance().registry(PropertyRegistry.class);
+
+	/**
+	 * Creates a new property. Prefer {@link #of(String, String, String, SkriptAddon, Class)}.
+	 *
+	 * @param name the name of the property
+	 * @param provider the addon that provides this property
+	 * @param handler the handler class for this property
+	 * @see #of(String, String, String, SkriptAddon, Class)
+	 */
+	public Property(String name, String description, String[] since, SkriptAddon provider, @NotNull Class<? extends Handler> handler) {
+		this.name = name.toLowerCase(Locale.ENGLISH);
+		this.description = description;
+		this.since = since;
+		this.provider = provider;
+		this.handler = handler;
+	}
+
+	/**
+	 * Gets a documentation-friendly ID for this property, based on its name.
+	 * May be overridden to provide a custom ID.
+	 *
+	 * @return a documentation-friendly ID for this property
+	 */
+	public String getDocumentationID() {
+		return name.replace(' ', '-').toLowerCase(Locale.ENGLISH);
+	}
+
+	/**
+	 * Helpful registration shortcut.
+	 * @return whether the property was successfully registered.
+	 */
+	public boolean register() {
+		return PROPERTY_REGISTRY.register(this);
+	}
+
+	/**
+	 * Creates a new property.
+	 *
+	 * @param name the name of the property
+	 * @param description a brief description of the property
+	 * @param provider the addon that provides this property
+	 * @param handler the handler class for this property
+	 * @param <HandlerClass> the type of the handler class
+	 * @param <Handler> the type of the handler
+	 * @return a new property
+	 */
+	public static <HandlerClass extends PropertyHandler<?>, Handler extends HandlerClass> Property<Handler> of(
```

Comment:
With the presence of documentation-related fields, I think a builder would be good.

Reply:
I tried one, just could not figure out how to get the generics to work. It would always give me `XYZHandler` instead of `XYZHandler<?, ?>`.